### DISCUSSION TITLE
How to export paired devices' S2 DSK PIN?

### DISCUSSION THREAD

**Discussion by TiddlyWiddly** [zwave-js/zwave-js#3910] [Q&A]
I have like 30 switches now and rather than keep around 30 pieces of cardboard with the PINs on them, how can I export the PINs for the devices that I've already setup?

**Comment by AlCalzone** [comment]
_**How**_ are you using zwave-js?

**Comment by TiddlyWiddly** [comment]
I’m using the HA supervisor core version. I tried swapping to the MQTT addon yesterday and didn’t see it there either so I rolled back.

**Reply by AlCalzone** [reply]
Zwavejs2mqtt has an export option in the SmartStart tab. It's hidden behind that action button in the corner.
How/if it works in the official HA addon, I don't know.

**Reply by TiddlyWiddly** [reply]
I followed instructions to migrate over from the official addon to the mqtt addon and even though all my switches work and are listed in the control panel of the mqtt addon, the smart start tab is empty (as I would expect I guess). I also don't see an action button in the corner of the smart start tab. Am I doing something wrong?

![image](https://user-images.githubusercontent.com/10444452/146295671-7e6d3538-2424-4698-b8ee-51e9336121e5.png)

**Reply by AlCalzone** [reply]
@robertsLando any idea?

**Reply by robertsLando** [reply]
@AlCalzone That list contians all the entries I get from `driver.controller.getProvisioningEntries` so if them are not listed there it means that call has return an empty array (or maybe there is an error but we need a log to spot that in case)

**Reply by AlCalzone** [reply]
@robertsLando I meant this:
> I also don't see an action button in the corner of the smart start tab. Am I doing something wrong?

@TiddlyWiddly it could be that the cache isn't migrated when switching between the official HA integration and z2m. You'll have to ask in the HomeAssistant channels how to get to that information.

**Reply by robertsLando** [reply]
Thre should be a fab in the bottom right corner: 
![Schermata da 2021-12-16 09-03-18](https://user-images.githubusercontent.com/11502495/146331743-d65062f3-a4ab-4262-a561-121e11b487b6.png)

**Comment by TiddlyWiddly** [comment]
Ok in the plus sign there is an export, but it exports an empty array, even though all switches are working and listed in the control panel.

**Reply by AlCalzone** [reply]
Well, you don't have any provisioning info stored here, so there is nothing to export.

If you can somehow get to the cache files in HA (I'm still unsure if they are synchronized between the integrations), you can download the `<homeid>.json` file and copy the nodes' `dsk` entries (if they are there). The PIN is just the first 5 digits.

If not, open a feature request in http://github.com/homeassistant/core (for the official integration) and/or https://github.com/zwave-js/zwavejs2mqtt (for the addon).

**Comment by schmitzkr** [comment]
Does anyone have a solution for this 2 years later? or do we know if a feature request was opened anywhere?
Thanks!

**Reply by robertsLando** [reply]
If you click on the floating action button on bottom right corner there are both the import/expèort actions to do that


![Peek 25-09-2023 11-45](https://github.com/zwave-js/node-zwave-js/assets/11502495/e5a5eb11-9c30-4de6-b828-945c611f669a)

**Reply by AlCalzone** [reply]
@robertsLando that's not what's being asked here I think. Nodes that have been included via S2, but not SmartStart also have a DSK and might require you to enter the PIN.
There's currently no way to export an overview which node has which DSK.

**Reply by robertsLando** [reply]
@AlCalzone Oh ok so I misunderstood the question sorry

**Reply by schmitzkr** [reply]
Thanks for your reply, I was able to get what I need calling the driver with a snippet.  Described here 
https://github.com/zwave-js/zwave-js-ui/issues/3296

**Comment by madbrain76** [comment]
@schmitzkr ,
I have the same need. What program did you use exactly with that "driver function" screen ? The default example fails.

**Comment by madbrain76** [comment]
Courtesy of a couple hours with ChatGPT . I don't know Javascript, so use at your own risk. But it worked for me.

```
// Function to export all Z-Wave Device Specific Keys (DSKs) with labels
async function exportAllDSKsWithLabels(driver) {
    // Check if the driver instance is valid
    if (!driver) {
        throw new Error("Driver instance is not available");
    }

    // Retrieve all nodes
    const nodes = driver.controller.nodes;

    // Prepare an array to hold DSKs with labels
    const dskList = [];

    // Iterate through all nodes to get their DSKs with labels
    for (const [nodeId, node] of nodes) {
        if (node.dsk) {
            // Convert DSK to 8 groups of 5 digits separated by hyphens
            const formattedDSK = formatDSK(node.dsk);
            const name = node.name || "Unknown";
            const location = node.location || "Unknown";
            const label = node.label || "Unknown";
            dskList.push({ nodeId, name, location, dsk: formattedDSK, label });
        }
    }

    // Format the DSKs with labels into a CSV string
    const csvHeader = "Node ID,Name,Location,DSK,Label\n";
    const csvRows = dskList.map(item => `${item.nodeId},${item.name},${item.location},${item.dsk},${item.label}`).join("\n");
    const csvContent = csvHeader + csvRows;

    // Log the CSV content to the console (or save to a file)
    console.log(csvContent);

    // Optionally, you can save the CSV content to a file using Node.js fs module
    // const fs = require('fs');
    // fs.writeFileSync('zwave_dsk_export_with_labels.csv', csvContent);

    return csvContent;
}

// Run the export function with the driver instance
exportAllDSKsWithLabels(driver)
    .then(csvContent => console.log("DSK export completed with labels"))
    .catch(err => console.error("Error exporting DSKs with labels:", err));

// Function to format DSK as 8 groups of 5 digits separated by hyphens
function formatDSK(dsk) {
    // Split the 16-byte DSK into 8 2-byte values
    const dskValues = [];
    for (let i = 0; i < 16; i += 2) {
        const value = dsk.readUInt16BE(i);
        dskValues.push(value);
    }

    // Convert each 2-byte value to a 5-digit decimal format
    const formattedValues = dskValues.map(value => value.toString().padStart(5, '0'));

    // Join the formatted values with hyphens
    return formattedValues.join('-');
}

```

**Reply by schmitzkr** [reply]
Sorry I don't recall how I ended up doing this. 
If that snippet works, I'm happy for you. If you still want to know then I can try to retrace my steps.

**Reply by AlCalzone** [reply]
@madbrain76 looks good to me, just two things I'd note.

`driver` should always be defined in the driver function, so there's no need to check it.
Also, Z-Wave JS exports a function to format the DSK as a string, so you could use this instead of your `formatDSK` function:
```ts
const { dskToString } = this.require("@zwave-js/core");
```

**Reply by madbrain76** [reply]
@AlCalzone thanks ! It took me a while to reverse engineer the algorithm to go from 16-byte binary DSK to the 40 digit decimal string. I wish I had known about this function. But as I said, I don't know any Javascript, so doing that was easier for me.

I'm a C/C++ programmer. I have no idea what that line of code you typed in your comment means, or how to apply it to the above program, unfortunately. It looks like it might be the equivalent of some #include statement. That's as far as I understand, though. I just don't know the JS syntax.

Where is that function documented, BTW ? I was also looking to possibly export other fields, such as the 3 S2 properties, the S0 property, the manufacturer, and description. But the code generated by ChatGPT for fetching those didn't work - it didn't return any value for those fields.

**Reply by AlCalzone** [reply]
Sorry I missed this.

> I have no idea what that line of code you typed in your comment means, or how to apply it to the above program, unfortunately.

`require("module-name")` is more or less the Node.js equivalent of an `#include` statement. `const { dskToString }` only takes the `dskToString` method from the external module. `this.require` is necessary in Z-Wave JS UI instead of `require` because of how Z-UI sets up the execution context for you.

The function should be documented [here](https://zwave-js.github.io/node-zwave-js/#/api/utils), but like a lot of the utility functions it unfortunately isn't.

[`PlannedProvisioningEntry`](https://zwave-js.github.io/node-zwave-js/#/api/controller?id=provisionsmartstartnode) describes which other properties (may) exist on a provisioning entry.

**Reply by madbrain76** [reply]
Thanks for the JS lesson. I may still need to go through some kind of tutorial if I'm to produce any more JS code. ChatGPT only gets me so far, especially as the program size grows and it forgets half of everything said before.

**Reply by tamorgen** [reply]
> Courtesy of a couple hours with ChatGPT . I don't know Javascript, so use at your own risk. But it worked for me.


@madbrain76 ,
I used your script above, and it told me that it was successful.  Where do I find the results?

**Reply by madbrain76** [reply]
> > Courtesy of a couple hours with ChatGPT . I don't know Javascript, so use at your own risk. But it worked for me.
> 
> @madbrain76 , I used your script above, and it told me that it was successful. Where do I find the results?

I don't recall, but looking at the source code, it appears to log to the console. This might be your browser's Javascript console.

**Reply by AlCalzone** [reply]
The console is the context you're executing Z-Wave JS UI from. Either directly on the command line, or Docker logs if using Docker, or Addon logs if using a Home Assistant addon.

**Comment by madbrain76** [comment]
The other question in my mind is why all those manuallyed included devices don't automatically appear in the SmartStart screen, which would be much easier than having to export them and import them.

**Reply by AlCalzone** [reply]
@robertsLando I think this sounds good. We just need to make it obvious that those are not SmartStart entries

**Reply by madbrain76** [reply]
@AlCalzone Are there S2 Auth devices out in the wild that don't support Smart Start ? It looks like SmartStart requires S2 security.
But S2 auth was introduced before SmartStart. I believe all my S2 auth devices are also SmartStart. But I bought them all within the past 3 years.
There are a couple different approaches to differentiate the devices in the UI :

1) based on querying the device capabilities . Not sure if this is possible to query. If yes, I think this would be the preferred method
2) based on some device database
3) based on how they were originally added - if that information is available. Ie. through SmartStart or manual inclusion. This will be the least reliable probably, though. I added about half manually, and half through SmartStart.

**Reply by robertsLando** [reply]
@AlCalzone How can I distinguish a Smart Start entry and a manual created entry?

**Reply by AlCalzone** [reply]
> Are there S2 Auth devices out in the wild that don't support Smart Start ?

Pretty sure, yes.

> How can I distinguish a Smart Start entry and a manual created entry?

The SmartStart entry exists in the list of provisioning entries.

If the node was normally included, there is no matching entry with the node's DSK. Unfortunately this is also true if a node was included via SmartStart but Unauthenticated. I guess it doesn't matter in that case though.

**Reply by robertsLando** [reply]
@AlCalzone wondering if it could make sense to make the driver set this in a property of entries to distinguish them? Otherwise I can do it on my side

**Reply by AlCalzone** [reply]
There are no entries for nodes that haven't been included via SmartStart. I don't think adding a property on the driver side saves a lot of work on your end, since you have to filter anyways. This should do the trick

```ts
const entries = driver.controller.getProvisioningEntries();
const ssNodeIds = new Set(
	entries.filter((e) => typeof e.nodeId === "number").map((e) =>
		e.nodeId
	),
);
const nonSmartStartNodes = [...driver.controller.nodes.values()].filter((n) =>
	!ssNodeIds.has(n.id)
);
```
